### PR TITLE
Use matching against escaped URLs

### PR DESCRIPTION
**Pull request by klizhentas**
@r0mant @horkhe Please review this fixes problems with routing of URLs containing escaped slashes and will reduce amount of errors in logs

### CODE REVIEW

File: proxy.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		// In case if it's redirect error, try the request one more time, but with different URL
 		r.URL = e.URL
 		r.Host = e.URL.Host
+		r.RequestURI = e.URL.String()
```
Comment (r0mant) [inline proxy.go:43]: How is `e.URL.String()` different from the value which is already in `r.RequestURI` field? From the docs:

> RequestURI is the unmodified Request-URI of the
> Request-Line (RFC 2616, Section 5.1) as sent by the client
> to a server. Usually the URL field should be used instead.
> It is an error to set this field in an HTTP client request.

File: netutils/netutils.go [hunk 2]
```diff
@@ -20,6 +20,30 @@ func CopyUrl(in *url.URL) *url.URL {
...
+		return "", err
+	}
+	if u.Opaque != "" {
+		return u.Opaque, nil
+	}
+	// This is local URL, has no host, return as-is
+	if u.Host == "" {
+		return in, nil
+	}
+	vals := strings.SplitN(in, u.Host, 2)
```
Comment (r0mant) [inline netutils/netutils.go:36]: You're assuming that URL has a scheme defined, e.g. `www.google.com/abc` will not work, right?

There are probably other edge cases. Would not it be more reliable to parse the URL, nullify path and fragment and reconstruct it using `URL.RequestURI()` function rather that do by-hand parsing?

File: proxy.go (hunk 1 above)
Comment (horkhe) [inline proxy.go:43]: By hacking the **RequestURI** field we are breaking the contract of the standard **http.Request** type, so we better introduce another field (e.g. in Proxy) for that.

File: route/exproute/trie.go [hunk 3]
```diff
@@ -64,7 +65,10 @@ func (p *trie) match(r request.Request) location.Location {
 		return nil
 	}
 
-	path := r.GetHttpRequest().URL.Path
+	path, err := netutils.RawPath(r.GetHttpRequest().RequestURI)
```
Comment (r0mant) [inline route/exproute/trie.go:68]: To my question above: it looks like you could just do `netutils.RawPath(r.GetHttpRequest().URL.String())` here, without overriding `RequestURI` field.

Comment (klizhentas) [inline route/exproute/trie.go:68]: URL always contains unescaped values

File: proxy.go (hunk 1 above)
Comment (klizhentas) [inline proxy.go:43]: This is redirect, so effectively we are forming a new request here

File: netutils/netutils.go (hunk 2 above)
Comment (klizhentas) [inline netutils/netutils.go:36]: URL data structure does not contain original data anywhere unfortunately,  correct solution would be to rewrite URL parsing ourselves and use it instead, but that's too much work

File: route/exproute/trie.go (hunk 3 above)
Comment (r0mant) [inline route/exproute/trie.go:68]: Oh, looks like this line is unrelated to the `r.RequestURI = e.URL.String()` above. Then it makes sense.

File: netutils/netutils.go (hunk 2 above)
Comment (r0mant) [inline netutils/netutils.go:36]: I guess you can't use `URL.RequestURI()` here anyways because it will unescape the string when reconstructing it anyways.

**Comment by r0mant** [comment]
lgtm

**Comment by horkhe** [comment]
lgtm
